expression/builtin_string.go

```diff
@@ -3976,3 +3976,83 @@ func (b *builtinWeightStringSig) evalString(row chunk.Row) (string, bool, error)
 	}
 	return string(ctor.Key(str)), false, nil
 }
+
+type translateFunctionClass struct {
+	baseFunctionClass
+}
+
+// getFunction sets translate built-in function signature.
+// The syntax of translate in Oracle is 'TRANSLATE(expr, from_string, to_string)'.
+func (c *translateFunctionClass) getFunction(ctx sessionctx.Context, args []Expression) (builtinFunc, error) {
+	if err := c.verifyArgs(args); err != nil {
+		return nil, err
+	}
+	bf, err := newBaseBuiltinFuncWithTp(ctx, c.funcName, args, types.ETString, types.ETString, types.ETString, types.ETString)
+	if err != nil {
+		return nil, err
+	}
+	argType := args[0].GetType()
+	bf.tp.Flen = argType.Flen
+	SetBinFlagOrBinStr(argType, bf.tp)
+	sig := &builtinTranslateSig{bf}
```

Comment:
🤔 

the current implementation won't work correctly when the string is in BINARY encoding because of the `[]rune()` cast. 

```sql
select translate(x'C3A4C3ABC3AFC3B6C3BC', x'A4B6C3', x'4A6B3C');
-- Expected?: 0x3C4A3CAB3CAF3C6B3CBC
```

you'll need to introduce `builtinTranslateSig` and `builtinTranslateUTF8Sig`. see the implementation of `func (c *lpadFunctionClass) getFunction` near line 1999 for example.

something that won't happen in Oracle is handling of mixed-encoding `TRANSLATE()`.

```
0. TRANSLATE(utf8, utf8, utf8) = utf8
1. TRANSLATE(utf8, utf8, binary) = binary?
2. TRANSLATE(utf8, binary, utf8) = binary?
3. TRANSLATE(utf8, binary, binary) = binary?
4. TRANSLATE(binary, utf8, utf8) = binary
5. TRANSLATE(binary, utf8, binary) = binary
6. TRANSLATE(binary, binary, utf8) = binary
7. TRANSLATE(binary, binary, binary) = binary
```

i've got question in cases 1-3 because the closely related `REPLACE()` function in MySQL 8 behaves like this instead: the chosen collation entirely follows the first argument only.

```sql
mysql> select replace('ü', x'c3', x'c2');
ERROR 3854 (HY000): Cannot convert string '\xC3' from binary to utf8mb4

mysql> select replace(cast('ü' as binary), x'c3', x'c2');
0xC2BC
```

in TiDB `REPLACE()` behaves differently from MySQL here:

```sql
TiDB> select replace('ü', x'c3', x'c2');
0xC2BC
```

Reply:
Fixed. If any arguments is binary string, then we will process byte-by-byte instead of rune-by-rune.